File: 0120-triangle/0120-triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        if(triangle.size() <= 1) return triangle[0][0]; 
        int m = triangle.size(); 
        for(int i = 1; i < m; i++){
            int top = triangle[i-1].size()-1; 
            int bottom = triangle[i].size()-1; 
            triangle[i][0] = triangle[i-1][0] + triangle[i][0]; 
            triangle[i][bottom] = triangle[i-1][top] + triangle[i][bottom]; 
        }

        for(int i = 2; i < m; i++){
            int n = triangle[i].size(); 
            for(int j = 1; j < n - 1; j++){
                triangle[i][j] = min(triangle[i-1][j-1],triangle[i-1][j]) + triangle[i][j]; 
            }
        }

        // for(int i = 0; i < m; i++){
        //     int n = triangle[i].size(); 
        //     for(int j = 0; j < n; j++){
        //         cout << triangle[i][j] << ' '; 
        //     }
        //     cout << endl; 
        // }

        return *min_element(triangle[m-1].begin(), triangle[m-1].end()); 
    }
};

File: 0120-triangle/0120-triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle.size() == 1) return triangle.get(0).get(0); 
        for(int i = 1; i < triangle.size(); i++){
            for(int j = 0; j < triangle.get(i).size(); j++){
                int curr = triangle.get(i).get(j); 
                int left_min = j > 0 ? triangle.get(i-1).get(j-1) : 10001; 
                int right_min = j < triangle.get(i-1).size() ? triangle.get(i-1).get(j) : 10001; 
                //triangle.get(i).get(j) = Math.min(curr + left_min, curr + right_min); 
                triangle.get(i).set(j,Math.min(curr + left_min, curr + right_min)); 
            }
        }
        

        return Collections.min(triangle.get(triangle.size()-1)); 
    }
}